### PR TITLE
btl/base_am_rdma: fix handling of btl_send returning 1

### PR DESCRIPTION
**Pull request by wzamazon**
Active message RDMA uses btl_send to send the initial
request and RDMA response.

btl_send will return 0 when the descriptor has been
successfully queued for send, and will return 1 when
the desciptor has been successfully sent.

Currently, active message RDMA treats the return value
1 as an error, thus will either return the value
to caller, or retry the send.

This patch address the issue by correctly handling
the return value 1.

Signed-off-by: Wei Zhang <wzam@amazon.com>

### CODE REVIEW

File: opal/mca/btl/base/btl_base_am_rdma.c [hunk 1]
```diff
@@ -409,7 +410,19 @@ static inline int mca_btl_base_am_rdma_advance(mca_btl_base_module_t *btl,
     }
 
     if (send_descriptor) {
-        return btl->btl_send(btl, endpoint, descriptor, mca_btl_base_rdma_tag(hdr->type));
+        /* btl_send return OPAL_SUCCESS when that the descriptor has been queued for send, and
```
Comment (bwbarrett) [inline opal/mca/btl/base/btl_base_am_rdma.c:413]: You don't need this big of a comment for the semantics (now that the btl.h file has the documentation on return value).  I would add a note that we always set `MCA_BTL_DES_SEND_ALWAYS_CALLBACK`, so we can treat 0 and 1 the same.  Just something as simple as:

```
/* This descriptor will have MCA_BTL_DES_SEND_ALWAYS_CALLBACK set, so safe to treat 0 and 1 return codes the same */
```

Or you could replace that comment with the even better assert:
```
assert(0 != (descriptor->flags && MCA_BTL_DES_SEND_ALWAYS_CALLBACK));
```
that way, you've both documented the assumption and proven the assumption (in debug builds, anyway).

**Review by bwbarrett — CHANGES_REQUESTED** [review]

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 1 above)
Comment (wzamazon) [reply opal/mca/btl/base/btl_base_am_rdma.c]: Thanks! In new revision, I removed the excessive comments, and added the assertion.

File: opal/mca/btl/base/btl_base_am_rdma.c [hunk 2]
```diff
@@ -409,7 +410,13 @@ static inline int mca_btl_base_am_rdma_advance(mca_btl_base_module_t *btl,
     }
 
     if (send_descriptor) {
-        return btl->btl_send(btl, endpoint, descriptor, mca_btl_base_rdma_tag(hdr->type));
+        /* This descriptor was created with MCA_BTL_DES_SEND_ALWAYS_CALLBACK set, so safe to treat
```
Comment (bwbarrett) [inline opal/mca/btl/base/btl_base_am_rdma.c:413]: You don't need the comment.  The assert says the same thing :).

File: opal/mca/btl/base/btl_base_am_rdma.c [hunk 3]
```diff
@@ -409,7 +410,13 @@ static inline int mca_btl_base_am_rdma_advance(mca_btl_base_module_t *btl,
...
+         * OPAL_SUCCESS and 1 return codes the same */
+        assert(0 != (descriptor->des_flags && MCA_BTL_DES_SEND_ALWAYS_CALLBACK));
+        ret = btl->btl_send(btl, endpoint, descriptor, mca_btl_base_rdma_tag(hdr->type));
+        if (ret == 1)
```
Comment (bwbarrett) [inline opal/mca/btl/base/btl_base_am_rdma.c:417]: Sorry, I missed this in the previous iteration.  Open MPI *always* uses curly braces around conditional bodies.

**Review by bwbarrett — CHANGES_REQUESTED** [review]

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 2 above)
Comment (wzamazon) [reply opal/mca/btl/base/btl_base_am_rdma.c]: Thank you! removed in new revision

File: opal/mca/btl/base/btl_base_am_rdma.c (hunk 3 above)
Comment (wzamazon) [reply opal/mca/btl/base/btl_base_am_rdma.c]: Thank you! fixed in new revision

**Review by bwbarrett — APPROVED** [review]

**Comment by wzamazon** [comment]
The PR build checker failed because a git issue:
```
> git checkout -f ee62a31826851bc8f0e509a4f5b97eebb4e50ac8 # timeout=10
FATAL: Could not checkout ee62a31826851bc8f0e509a4f5b97eebb4e50ac8
hudson.plugins.git.GitException: Command "git checkout -f ee62a31826851bc8f0e509a4f5b97eebb4e50ac8" returned status code 128:
stdout: 
stderr: fatal: reference is not a tree: ee62a31826851bc8f0e509a4f5b97eebb4e50ac8
```
should not be related to the PR.

**Comment by bwbarrett** [comment]
bot:aws:retest

**Pull request merged**